### PR TITLE
use topology to achive HA

### PR DESCRIPTION
**Pull request by PengJi**
### What problem does this PR solve? 
#2110

### What is changed and how does it work?
Some label is set to node, for example: kubernetes.io/hostname=host1,zone=zone1, then a certain label is used as tolology, and the scheduler schedules pod to different tolology according to tolology to implement [HA](https://pingcap.com/docs-cn/tidb-in-kubernetes/stable/tidb-scheduler/)

### Check List 

Tests 

 - Unit test
 - E2E test
 - Manual test (add detailed scripts or steps below)

Code changes
Mainly changed ha.go

Side effects

NONE

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

### Does this PR introduce a user-facing change?:

```release-note
NONE
```

### CODE REVIEW

**Comment by Yisaer** [comment]
Thanks for your contribution. Could you provide more detail and some examples to demonstrate the scenario and how it works?

File: charts/tidb-cluster/templates/tidb-cluster.yaml [hunk 1]
```diff
@@ -7,6 +7,7 @@ metadata:
     pingcap.com/pd.{{ template "cluster.name" . }}-pd.sha: {{ include "pd-configmap.data-digest" . | quote }}
     pingcap.com/tikv.{{ template "cluster.name" . }}-tikv.sha: {{ include "tikv-configmap.data-digest" . | quote }}
     pingcap.com/tidb.{{ template "cluster.name" . }}-tidb.sha: {{ include "tidb-configmap.data-digest" . | quote }}
+    topologyKey: {{ .Values.topologyKey }}
```
Comment (weekface) [inline charts/tidb-cluster/templates/tidb-cluster.yaml:10]: What about change `topologyKey` to `pingcap.com/ha-topology-key`?

and define it as a const like https://github.com/pingcap/tidb-operator/blob/master/pkg/label/label.go#L43 does.

File: charts/tidb-cluster/values.yaml [hunk 2]
```diff
@@ -64,6 +64,10 @@ discovery:
 # if the ConfigMap was not changed.
 enableConfigMapRollout: true
 
+# The scheduler achieves HA scheduling based on the topologyKey.
+# You can modify it to other label of node
+topologyKey: kubernetes.io/hostname
```
Comment (weekface) [inline charts/tidb-cluster/values.yaml:69]:
```suggestion
haTopologyKey: kubernetes.io/hostname
```

**Comment by PengJi** [comment]
> Thanks for your contribution. Could you provide more detail and some examples to demonstrate the scenario and how it works?

The current tidb-operator's scheduling rule is highly available scheduling in the node dimension, which guarantees that the component instances are scheduled to different nodes to ensure high availability. Sometimes the distribution of nodes is as follows:

| zone | zone1 | zone2 | zone3 | zone3 | zone3 |
| -- | -- | -- | -- |  -- | -- | 
|node | kind-worker | kind-worker2 | kind-worker3 | kind-worker4 | kind-worker5
|pd |basic-pd-0 |   | basic-pd-1 | basic-pd-2 |  
| tikv |basic-tikv-0 | basic-tikv-1 |   |   | basic-tikv-2 |


The zones here are assumed to be different IDCs. It can be seen that kind-worker is located in zone1, kind-worker2 is located in zone2, kind-worker3, kind-worker4 and kind-worker5 are all located in zone3. At this time, basic-po-1 and basic-pd-2 are located in the same zone (zone3). We sometimes want to schedule pod to different zones, that is, to ensure the zone dimension is highly available, as follows:

| zone | zone1 | zone2 | zone3 | zone3 | zone3 |
| -- | -- | -- | -- |  -- | -- | 
|node | kind-worker | kind-worker2 | kind-worker3 | kind-worker4 | kind-worker5
|pd |basic-pd-0 | basic-pd-1 |   | basic-pd-2  | |  
| tikv |basic-tikv-0 | basic-tikv-1 |   |   | basic-tikv-2

This ensures that at least two pods are still available in case of zone3 power outage, network outage, etc.

The code implements a highly available schedule for the zone dimension, where the zone, if not configured (I'm currently changing it to be configured through tidb-cluster.yaml annotations), defaults to the node, which is compatible with the current scheduling rules of tidb-operator

**Comment by Yisaer** [comment]
@PengJi  Thanks for clarifying. That's very helpful.

**Comment by Yisaer** [comment]
Plz gofmt the `pkg/scheduler/predicates/test_helper.go` and `pkg/scheduler/predicates/predicate_test.go`

**Comment by DanielZhangQD** [comment]
@PengJi Thanks for the contribution!
Can this be done via [pod anti-affinity](https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#inter-pod-affinity-and-anti-affinity)? @PengJi @weekface

**Comment by PengJi** [comment]
> @PengJi Thanks for the contribution!
> Can this be done via [pod anti-affinity](https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#inter-pod-affinity-and-anti-affinity)? @PengJi @weekface

@DanielZhangQD  One of our requirements is to implement HA based on the zone dimension (such as IDC, currently is node dimension). If using podAntiAffinity, the possible configuration is as follows:
```
nodeAffinity:
  requiredDuringSchedulingIgnoredDuringExecution:
    nodeSelectorTerms:
    - matchExpressions:
      - key: failure-domain.beta.kubernetes.io/zone
        operator: In
        values:
        - zone1
        - zone2
        - zone3
podAntiAffinity:
  preferredDuringSchedulingIgnoredDuringExecution:
  - weight: 100
    podAffinityTerm:
      labelSelector:
        matchExpressions:
        - key: app.kubernetes.io/component
          operator: In
          values:
          - pd
      topologyKey: failure-domain.beta.kubernetes.io/zone
```

This method does not guarantee that pods are evenly distributed among the three different zones, or is there any other configuration method?
In fact, it is similar to [TopologySpreadConstraints](https://kubernetes.io/docs/concepts/workloads/pods/pod-topology-spread-constraints/), but our k8s version is relatively low (< 1.18), cannot use this feature in our production environment.

File: charts/tidb-cluster/templates/tidb-cluster.yaml (hunk 1 above)
Comment (weekface) [inline charts/tidb-cluster/templates/tidb-cluster.yaml:10]:
```suggestion
    topologyKey: {{ .Values.topologyKey | default "kubernetes.io/hostname" }}
```

**Comment by weekface** [comment]
@cofyc PTAL

**Comment by weekface** [comment]
@PengJi Can you add a release note (user-facing change)?

File: charts/tidb-cluster/templates/tidb-cluster.yaml (hunk 1 above)
Comment (PengJi) [reply charts/tidb-cluster/templates/tidb-cluster.yaml]: updated

Comment (PengJi) [reply charts/tidb-cluster/templates/tidb-cluster.yaml]: updated

File: charts/tidb-cluster/values.yaml (hunk 2 above)
Comment (PengJi) [reply charts/tidb-cluster/values.yaml]: updated

**Comment by cofyc** [comment]
> > Thanks for your contribution. Could you provide more detail and some examples to demonstrate the scenario and how it works?
> 
> The current tidb-operator's scheduling rule is highly available scheduling in the node dimension, which guarantees that the component instances are scheduled to different nodes to ensure high availability. Sometimes the distribution of nodes is as follows:
> 
> zone	zone1	zone2	zone3	zone3	zone3
> node	kind-worker	kind-worker2	kind-worker3	kind-worker4	kind-worker5
> pd	basic-pd-0	 	basic-pd-1	basic-pd-2	 
> tikv	basic-tikv-0	basic-tikv-1	 	 	basic-tikv-2
> The zones here are assumed to be different IDCs. It can be seen that kind-worker is located in zone1, kind-worker2 is located in zone2, kind-worker3, kind-worker4 and kind-worker5 are all located in zone3. At this time, basic-po-1 and basic-pd-2 are located in the same zone (zone3). We sometimes want to schedule pod to different zones, that is, to ensure the zone dimension is highly available, as follows:
> 
> zone	zone1	zone2	zone3	zone3	zone3
> node	kind-worker	kind-worker2	kind-worker3	kind-worker4	kind-worker5
> pd	basic-pd-0	basic-pd-1	 	basic-pd-2	
> tikv	basic-tikv-0	basic-tikv-1	 	 	basic-tikv-2
> This ensures that at least two pods are still available in case of zone3 power outage, network outage, etc.
> 
> The code implements a highly available schedule for the zone dimension, where the zone, if not configured (I'm currently changing it to be configured through tidb-cluster.yaml annotations), defaults to the node, which is compatible with the current scheduling rules of tidb-operator

This LGTM and is an enhancement to current HA implementation.

It does not support more than one topology, so pods in a particular zone (topology-key=zone) will not be spread evenly across nodes. Do you plan to implement this feature, like [Multiple TopologySpreadConstraints](https://kubernetes.io/docs/concepts/workloads/pods/pod-topology-spread-constraints/#example-multiple-topologyspreadconstraints)?

**Comment by PengJi** [comment]
> > > Thanks for your contribution. Could you provide more detail and some examples to demonstrate the scenario and how it works?
> > 
> > 
> > The current tidb-operator's scheduling rule is highly available scheduling in the node dimension, which guarantees that the component instances are scheduled to different nodes to ensure high availability. Sometimes the distribution of nodes is as follows:
> > zone	zone1	zone2	zone3	zone3	zone3
> > node	kind-worker	kind-worker2	kind-worker3	kind-worker4	kind-worker5
> > pd	basic-pd-0	 	basic-pd-1	basic-pd-2	 
> > tikv	basic-tikv-0	basic-tikv-1	 	 	basic-tikv-2
> > The zones here are assumed to be different IDCs. It can be seen that kind-worker is located in zone1, kind-worker2 is located in zone2, kind-worker3, kind-worker4 and kind-worker5 are all located in zone3. At this time, basic-po-1 and basic-pd-2 are located in the same zone (zone3). We sometimes want to schedule pod to different zones, that is, to ensure the zone dimension is highly available, as follows:
> > zone	zone1	zone2	zone3	zone3	zone3
> > node	kind-worker	kind-worker2	kind-worker3	kind-worker4	kind-worker5
> > pd	basic-pd-0	basic-pd-1	 	basic-pd-2
> > tikv	basic-tikv-0	basic-tikv-1	 	 	basic-tikv-2
> > This ensures that at least two pods are still available in case of zone3 power outage, network outage, etc.
> > The code implements a highly available schedule for the zone dimension, where the zone, if not configured (I'm currently changing it to be configured through tidb-cluster.yaml annotations), defaults to the node, which is compatible with the current scheduling rules of tidb-operator
> 
> This LGTM and is an enhancement to current HA implementation.
> 
> It does not support more than one topology, so pods in a particular zone (topology-key=zone) will not be spread evenly across nodes. Do you plan to implement this feature, like [Multiple TopologySpreadConstraints](https://kubernetes.io/docs/concepts/workloads/pods/pod-topology-spread-constraints/#example-multiple-topologyspreadconstraints)?

There is no plan to implement this feature temporarily. The current method is to use the k8s SelectorSpreadPriority and InterPodAffinityPriority priority strategy and increase their weight.

**Pull request merged**